lib/Sema/CSApply.cpp

```diff
@@ -3629,13 +3632,63 @@ namespace {
       llvm_unreachable("Unhandled OverloadChoiceKind in switch.");
     }
 
+    /// Builds an argument list for making a call to the given
+    /// `@dynamicMemberLookup` subscript by inserting any necessary default
+    /// arguments.
+    ArgumentList *buildArgumentListForDynamicMemberLookupSubscript(
+        const SelectedOverload &overload, SourceLoc componentLoc,
+        ConstraintLocator *locator) {
+      assert(overload.choice.isAnyDynamicMemberLookup() &&
+             "Overload must be for dynamic member lookup");
+      auto *SD = cast<SubscriptDecl>(overload.choice.getDecl());
+      auto memberLoc = cs.getCalleeLocator(locator);
+      auto subscript = resolveConcreteDeclRef(SD, memberLoc);
+
+      // We specifically don't want the full type here to avoid losing parameter
+      // pack references.
+      auto fnType = overload.adjustedOpenedType->castTo<FunctionType>();
+
+      auto params = fnType->getParams();
+      SmallVector<Expr *, 4> args;
+      for (auto paramIdx : indices(params)) {
+        auto param = params[paramIdx];
+        auto paramTy =
+            simplifyType(param.getParameterType(/*forCanonical=*/true, &ctx));
+
+        Expr *argExpr;
+        if (paramIdx == 0) {
+          if (overload.choice.isKeyPathDynamicMemberLookup()) {
+            argExpr = buildKeyPathDynamicMemberArgExpr(paramTy, componentLoc,
+                                                       memberLoc);
+          } else {
+            auto fieldName =
+                overload.choice.getName().getBaseIdentifier().str();
+            argExpr = buildDynamicMemberLookupArgExpr(fieldName, componentLoc,
+                                                      paramTy);
+          }
+        } else {
+          argExpr = new (ctx) DefaultArgumentExpr(subscript, paramIdx,
+                                                  componentLoc, paramTy, dc);
+        }
+
+        if (!argExpr) {
+          return nullptr;
+        }
+
+        args.emplace_back(cs.cacheType(argExpr));
+      }
+
+      return ArgumentList::forImplicitCallTo(SD->getIndices(), args, ctx);
+    }
+
     /// Form a type checked expression for the argument of a
     /// @dynamicMemberLookup subscript index parameter.
     Expr *buildDynamicMemberLookupArgExpr(StringRef name, SourceLoc loc,
                                           Type literalTy) {
       // Build and type check the string literal index value to the specific
       // string type expected by the subscript.
       auto *nameExpr = new (ctx) StringLiteralExpr(name, loc, /*implicit*/true);
+      nameExpr->setType(literalTy);
```

Comment:
This shouldn't be necessary, during CSApply the ConstraintSystem is the source of truth for the types of the AST nodes, we then go through and set the types on the nodes when we're done (I know, it's weird, but it would be good to stick to the convention I think).